[gdscript/en] Update for Godot 4 (#4607)
adambard/learnxinyminutes-docs: gdscript2.html.markdown

---
language: GDScript 2.0
contributors:
    - ["Wichamir", "https://github.com/Wichamir/"]
    - ["zacryol", "https://github.com/zacryol"]
filename: learngdscript2.gd
---

> GDScript 2.0 is a dynamically and statically typed scripting language
> made specifically for free and open source game engine Godot, now revamped
> for the new Godot Engine 4.0. GDScript's syntax is vaguely similar to
> Python's. Its main advantages are ease of use and tight integration with
> the engine. It's a perfect fit for game development.

Reviewer: ```suggestion
```
Shorter is better.

## Basics

```nim
# Single-line comments are written using hash symbol.
"""
  Multi-line
  comments
  are
  written
  using
  triple
  quoted
  strings
"""

# Doc Comments can add a decription to classes and fields
# which can be viewed in the in-engine docs.

## This class is a demonstration of GDScript 2.0

# Script file is a class in itself and you can optionally define a name for it.
class_name MyClass

# Inheritance
extends Node2D

# Member variables
var x = 8 # int
var y = 1.2 # float
var b = true # bool
var s = "Hello World!" # String

> var a = [1, false, "brown fox"] # Array - similar to list in Python,
>                                 # it can hold different types
>                                 # of variables at once.

Reviewer: ```suggestion
var a = [1, false, "brown fox"] # Arrays hold different types of variables at once.
```

  return first + second

# Doc Comments on functions

## Increases the Jump Count
func jump():
  jump_count += 1

# Printing values
func printing():
  print("GDScript ", "is ", " awesome.")
  prints("These", "words", "are", "divided", "by", "spaces.")
  printt("These", "words", "are", "divided", "by", "tabs.")
  printraw("This gets printed to system console.")

  # Lambdas
  var my_lambda = func(): print("hello from lambda!")

  my_lambda.call()

# Math
func doing_math():
  var first = 8
  var second = 4
  print(first + second) # 12
  print(first - second) # 4
  print(first * second) # 32
  print(first / second) # 2
  print(first % second) # 0
  # There are also +=, -=, *=, /=, %= etc.,
  # however no ++ or -- operators.
  print(pow(first, 2)) # 64
  print(sqrt(second)) # 2
  printt(PI, TAU, INF, NAN) # built-in constants

# Control flow
func control_flow():
  x = 8
  y = 2 # y was originally a float,
        # but we can change its type to int
        # using the power of dynamic typing!
  
  if x < y:
    print("x is smaller than y")
  elif x > y:
    print("x is bigger than y")
  else:
    print("x and y are equal")
  
  var a = true
  var b = false
  var c = false

>   if a and b or not c: # alternatively you can use &&, || and !

Reviewer: ```suggestion
  if a and b or not c: # alternatively use &&, || and !
```

    print("This is true!")

>   for i in range(20): # GDScript's range is similar to Python's
>     print(i) # so this will print numbers from 0 to 19

Reviewer: ```suggestion
  for i in range(20): # prints numbers from 0 to 19
```

>   for i in 20: # unlike Python, you can loop over an int directly

Reviewer: ```suggestion
  for i in 20: # you can loop over an int directly
```

>     print(i) # so this will also print numbers from 0 to 19

Reviewer: ```suggestion
    print(i) # will print numbers from 0 to 19
```
